### ISSUE TITLE
prefer-const throws error even though let-defined variable is assigned multiple times.

### ISSUE THREAD

**Issue by indexzero** [eslint/eslint#5555]
- **What version of ESLint are you using?** `2.2.0`
- **What configuration and parser (Espree, Babel-ESLint, etc.) are you using?** `Espree`
- **What did you do? Please include the actual source code causing the issue.** See [the simple as possible reproduction](https://gist.github.com/indexzero/0df4264d5a1121469d2e)
- **What did you expect to happen?** No lint errors
- **What actually happened? Please include the actual, raw output from ESLint.** the `prefer-const` rule is violated even though the variable with `let` is clearly defined multiple times [as demonstrated by the output of the code execution](https://gist.github.com/indexzero/0df4264d5a1121469d2e#file-output-log)

**Comment by BYK** [comment]
You are not assigning the variable multiple times in your example, you just do it once so the rule is behaving correctly. That said the late binding issue was brought up before so I consider this as a duplicate of #5284 and closing this one. Feel free to reopen and provide more details if you disagree.

PS: We'd prefer if you can include your configuration and examples inside the bug report using code-fences since using an external page (even if it is a nice and tidy gist as you did) makes it harder to quickly evaluate the issue, especially on mobile devices).

**Comment by indexzero** [comment]
@BYK I don't quite understand the first part of response since the variable is assigned to twice. The program output verifies this. Apologies for not including it in the issue itself, but happy to walk through it again. Using this `eslintconfig.json`

``` js
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module",
  },
  "rules": {
    "prefer-const": 2
  }
}
```

We execute the following code:

``` js
// 1. Require `assert` and the variables define variables
const assert = require('assert');

let first;
let second;
let third;

// 2. Define a function that, when invoked assigns to all of the 
// variables we've defined (first, second, third). Output the value of
// third so that we can show explicitly it holds two distinct
// values over the run-time of the program.
function setupTest() {
  first = Math.random();
  second = Math.random();
  third = Math.random();
  console.log('third = %s', third);
}

// 3. We invoke setupTest twice. This leads to multiple assignments of each 
// of these variables. They are consumed outside the scope of "setupTest" so the
// placement of the `let` statements themselves is necessary.
setupTest();
assert.equal(typeof first, 'number');
assert.equal(typeof second, 'number');

setupTest();
assert.equal(typeof first, 'number');
assert.equal(typeof second, 'number');
```

Running this program we see that `third` does, in fact, hold two different values over the run-time of the program it is not possible for it to be a constant.

```
$ node test/prefer-const.js 
third = 0.10418695327825844
third = 0.8143592826090753
```

All this being said it is definitely a duplicate of #5284 because `prefer-const` is essentially _"detecting it can be moved to a smaller scope"_ in this case. For example I can remove the error by adding any additional reference to `third` outside of the scope of `setupTest` (not a reference with an assignment to `third`, just a reference).

**Comment by BYK** [comment]
>@BYK I don't quite understand the first part of response since the variable is assigned to twice. 

Oh, right! Sorry I skimmed the code and didn't really evaluate it in my head. Yeah it is definitely being assigned twice because of the two `setupTest()` calls. Apologies for the confusion :)

**Comment by ilyavolodin** [comment]
Looks like the issue should be reopened.

**Comment by BYK** [comment]
@mysticatea - any thoughts about this? I think #5284 would solve this since there's no easy way to determine multiple assignments done through multiple invocations of a function, right?

**Comment by mysticatea** [comment]
Thank you for this issue.

Yes, #5284 would solve this issue. After #5284, `prefer-const` will get not warning a variable if the variable moves to another scope.
Actually, I have done to implement #5284. I'm writing tests and docs now.

**Comment by BYK** [comment]
@mysticatea @ilyavolodin then I think we should close this issue.

**Comment by ilyavolodin** [comment]
Closing per discussion. This will be fixed as part of the #5284 issue.
